### PR TITLE
workflows/tests: use the container and `macos-14` for some jobs

### PR DESCRIPTION
**Pull request by ZhongRuoyu**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

The containers have `homebrew/core` tapped so we don't need to wait forever for the core repo to be cloned.

The tap syntax job requires all official taps. Unlike the Ubuntu images [^1], `macos-14` has both core and cask tapped [^2] so setting up Homebrew should be faster there. We don't use the containers because the cask tap is not available by default.

[^1]: https://github.com/actions/runner-images/blob/ubuntu24/20240922.1/images/ubuntu/scripts/build/install-homebrew.sh
[^2]: https://github.com/actions/runner-images/blob/macos-14/20240923.101/images/macos/scripts/build/install-homebrew.sh#L23-L24

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -242,21 +244,30 @@ jobs:
...
-            runs-on: ubuntu-22.04
+            runs-on: ubuntu-latest
+            container:
+              image: ghcr.io/homebrew/ubuntu22.04:master
           - name: tests (generic OS)
             test-flags: --generic --coverage
-            runs-on: ubuntu-22.04
+            runs-on: ubuntu-latest
+            container:
+              image: ghcr.io/homebrew/ubuntu22.04:master
```
Comment (MikeMcQuaid) [inline .github/workflows/tests.yml:260]:
```suggestion
              image: ghcr.io/homebrew/brew:master
```

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -242,21 +244,30 @@ jobs:
...
+    container: ${{ matrix.container }}
     strategy:
       matrix:
         include:
           - name: tests (online)
             test-flags: --online --coverage
-            runs-on: ubuntu-22.04
+            runs-on: ubuntu-latest
+            container:
+              image: ghcr.io/homebrew/ubuntu22.04:master
```
Comment (MikeMcQuaid) [inline .github/workflows/tests.yml:255]:
```suggestion
              image: ghcr.io/homebrew/brew:master
```

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -138,7 +138,9 @@ jobs:
     name: formula audit
     needs: syntax
     if: github.repository_owner == 'Homebrew'
-    runs-on: ubuntu-22.04
+    runs-on: ubuntu-latest
+    container:
+      image: ghcr.io/homebrew/ubuntu22.04:master
```
Comment (MikeMcQuaid) [inline .github/workflows/tests.yml:143]:
```suggestion
      image: ghcr.io/homebrew/brew:master
```

**Review by MikeMcQuaid — APPROVED** [review]
Nice work @ZhongRuoyu! Makes sense to me. Have suggested tweaks to defaults for a few containers where we don't explicitly mention the version to avoid needing updates in future and potentially catch issues quicker on the default container.

**Comment by MikeMcQuaid** [comment]
> We don't have `homebrew/brew:master` though, only `homebrew/ubuntu22.04` for `master`:

Ooo, my bad.

> Perhaps we can publish `homebrew/brew:master` too?

Yeh, I think it'd be ideal to have a `master` Docker build of some sort that can be tracked without caring about the underlying Ubuntu version?

That doesn't need to block this PR, though!

**Comment by ZhongRuoyu** [comment]
> Yeh, I think it'd be ideal to have a `master` Docker build of some sort that can be tracked without caring about the underlying Ubuntu version?

Agreed. Here we go: #18396

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by ZhongRuoyu** [comment]
The codecov action [failed](https://github.com/Homebrew/brew/actions/runs/11015185551/job/30587676343?pr=18395#step:14:25) because it's trying to [download a binary to the action's directory](https://github.com/codecov/codecov-action/blob/e28ff129e5465c2c0dcc6f003fc735cb6ae0c673/src/index.ts#L44-L54):

`Error: Codecov: Failed to write uploader binary: EACCES: permission denied, open '/__w/_actions/codecov/codecov-action/e28ff129e5465c2c0dcc6f003fc735cb6ae0c673/dist/codecov'`

Is it possible to set up ACL for the actions directory in setup-homebrew, @Bo98? Though there doesn't seem to be an environment variable for that.

**Comment by Bo98** [comment]
Maybe - I'll need to investigate to see.

Quick initial question though: do we get any speedup from using the containers for RSpec tests? Homebrew/core shouldn't be required there.

**Comment by ZhongRuoyu** [comment]
Not sure whether core is required for them, but:

> do we get any speedup from using the containers for RSpec tests?

The speedup is only in not having to tap `homebrew/core`. That aside, the speed is roughly the same (see [before](https://github.com/Homebrew/brew/actions/runs/11014747095/job/30586170103) and [after](https://github.com/Homebrew/brew/actions/runs/11015185551/job/30587676343)).

**Comment by ZhongRuoyu** [comment]
> Homebrew/core shouldn't be required there.

Is https://github.com/Homebrew/brew/pull/15634#discussion_r1253480618 no longer the case?

**Comment by Bo98** [comment]
Ah there's two remaining tests that still use Homebrew/core: https://github.com/search?q=repo%3AHomebrew%2Fbrew%20needs_homebrew_core&type=code.

Though maybe the macOS coverage is good enough for them.

**Review by Bo98 — APPROVED** [review]

**Pull request merged**